src/main/java/com/eerussianguy/blazemap/mixin/ServerLevelMixin.java

```diff
@@ -1,26 +1,44 @@
 package com.eerussianguy.blazemap.mixin;
 
+import java.util.List;
+import java.util.concurrent.Executor;
+
 import net.minecraft.core.BlockPos;
+import net.minecraft.core.Holder;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.progress.ChunkProgressListener;
+import net.minecraft.world.level.Level;
 import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.CustomSpawner;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.chunk.ChunkGenerator;
+import net.minecraft.world.level.dimension.DimensionType;
+import net.minecraft.world.level.storage.LevelStorageSource;
+import net.minecraft.world.level.storage.ServerLevelData;
 
 import com.eerussianguy.blazemap.engine.server.BlazeMapServerEngine;
+
 import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
 
 @Mixin(ServerLevel.class)
-public abstract class ServerLevelMixin {
+public abstract class ServerLevelMixin extends Level {
+    // This won't be called because it's a mixin, but Java still wants it to exist at compile time,
+    // so directly copying from ServerLevel
+    public ServerLevelMixin(MinecraftServer p_203762_, Executor p_203763_, LevelStorageSource.LevelStorageAccess p_203764_, ServerLevelData p_203765_, ResourceKey<Level> p_203766_, Holder<DimensionType> p_203767_, ChunkProgressListener p_203768_, ChunkGenerator p_203769_, boolean p_203770_, long p_203771_, List<CustomSpawner> p_203772_, boolean p_203773_) {
+      super(p_203765_, p_203766_, p_203767_, p_203762_::getProfiler, false, p_203770_, p_203771_);
+    }
 
-    @Inject(method = "onBlockStateChange", at=@At("HEAD"))
-    public void onBlockStateChange(BlockPos block, BlockState s1, BlockState s2, CallbackInfo ci) {
-        // If it's stupid but it works, it's not stupid.
-        // @Shadow fucks with the project because it doesn't reobf properly.
-        // No one knows why, no one can / wants to help, and so no one can point fingers at us either.
-        ServerLevel level = (ServerLevel) (Object) this;
+    @Shadow
+    public abstract void onBlockStateChange(BlockPos block, BlockState s1, BlockState s2);
 
-        BlazeMapServerEngine.onChunkChanged(level.dimension(), new ChunkPos(block));
+    @Inject(method = "onBlockStateChange", at = @At("HEAD"), expect = 1)
+    public void onOnBlockStateChange(BlockPos block, BlockState s1, BlockState s2, CallbackInfo ci) {
+        BlazeMapServerEngine.onChunkChanged(super.dimension(), new ChunkPos(block));
```

Comment:
Why `super` instead of `this`?
There might be some Mixin magic here, but this would most likely call the `dimension()` method in the target's (ServerLevel) superclass (Level) so if the target were to at some point override the method from the parent we wouldn't be calling the local method but the inherited one.
This should be fine for this one very specific instance but I'm still not a huge fan of how it could break with no warning.

Reply:
Let me take your second point first. Because this function call operates as if it's within `ServerLevel`, not the non-existent-at-runtime `ServerLevelMixin`, it can call functions the same way they could be called from within `ServerLevel`. As `.dimension()` is defined on the super class `Level`, not on `ServerLevel`, I can access it using `super`.

As for your first point, why `super` and not `this`, it's mainly because I was just throwing everything at the wall to get it to work in the first place and that seemed an easy way to get access to the function without hacking around, since at compile time `ServerLevelMixin` will recognise the existence of `.dimension()` on its super class, `Level`, which will be the same function once it's mixed into `ServerLevel`. 

It was around 2am by the time I was tidying up the changes (removing debug statements and commented out code) and didn't have the braincells left to mess with it further to see if I could further simplify things. Looking at it again with an awake brain and more confidence in `@Shadow`ing, I think I should be able to `@Shadow` the `.dimension()` as inherited by `ServerLevel` and call that directly, or even straight up access `this.dimension` to save an additional indirection. If I've the time after responding to comments here, might do a quick further tidy to fix stuff like this and push an update, otherwise will do tonight.

Reply:
Since we have direct access to the field then I'd just go for it instead of the method, and drop the shadow method.
This simplifies stuff and makes the this vs super argument moot.